blueprint/core/blueprint_ReactApplicationRoot.h

```diff
@@ -74,7 +74,13 @@ namespace blueprint
                 {
                     const auto lmt = wb.bundle.getLastModificationTime();
 
-                    if (lmt > wb.bundleLastModifiedTime)
+                    // In some instances webpack rebuilds temporarily result in an empty
+                    // bundle file before rebuilding. We do not want to trigger evaluation
+                    // callbacks when the bundle file is empty as this is likely to cause errors
+                    // when users attempt to call javascript functions from C/C++. Calling eval/call
+                    // in duktape with an empty js file/string does not result in an error and simply
+                    // pushes undefined to the top of the duktape stack as the
+                    if (lmt > wb.bundleLastModifiedTime && wb.bundle.loadFileAsString().isNotEmpty())
```

Comment:
This should probably have been in a separate commit but I'll squash this all down into one when we merge.

Reply:
Totally happy to have this here! I'll reply here also to your comment above about this saga since it seems a good place for it :D First, WOW. Thank you x1000 for going through that, what a pain.

>     1. Do we add checks for empty strings/bundle files in the two `EcmascriptEngine::evaluate` calls and assert and/or bail out?

Maybe I'm misunderstanding this slightly, but evaluating an empty string should always be a valid operation– an empty string is a valid expression. So for the `evaluate(const juce::String&)` overload, I think we should allow an empty string argument and make sure we don't call `duk_require_function`. If Duktape then has trouble with it, that seems like a Duktape bug (if that's the case, I have a workaround, but hope to not use it). Then for the `evaluate(const juce::File&)` overload, I'm wondering if this is a bit of a gotcha with `duk_compile_file`– maybe `duk_pcompile_string_filename` on an empty string doesn't leave a function on top? Maybe in that case, for the `evaluate(const juce::File&)` overload we should check if `file.isEmpty()` and just return early if it is?

>     2. With `EcmascriptEngine::invoke` should we only call `duk_require_function` when JUCE_DEBUG is not enabled? There is every chance we might remove some JS function whilst hot reloading that C++ code attempts to call. Alternatively, rather than removing in release, it might make sense to stick a debug print or comment next to the `duk_require_function` call, informing users that they have attempted to call a JS function which doesn't exist if it throws.

Ahh ok, so during the bundle reload, something like `CanvasView` might try to call `props["onDraw"]` which will have been wiped out already in the js environment? Originally I liked the idea of `duk_require_function` in the `invoke` implementation because if you invoke something that doesn't exist, it's a problem. But, now that I'm thinking about it again, attempting to invoke a non-existant function _is_ valid javascript, it should just throw a js error. With that in mind, I'm in favor of removing the `duk_require_function` line altogether, and then this discussion merges with the "how to handle errors the right way" discussion from below

Reply:
Yeah duktape handles calling evaluate with empty strings just fine. I'll sanity check when the undefined object was pushed to the stack after evaluating an empty string, I thiiiink it was at the point of `duk_pcall` inside `EcmascriptEngine::evaluate`, i.e. after the compile. I'll sanity check that.  

Cool with removing `duk_require_function` the alternative I was going to suggest was only call it if `JUCE_DEBUG` is not set. 

An early return on an empty file sounds fine to me. Do we put this at the `EcmascriptEngine::evaluate` level? In theory we could just remove the change to the bundle watcher than but maybe it's worth leaving that too to avoid the needless work of the `handleBundleChanged` callback?

Reply:
I like keeping the check in there on the bundle watcher, that feels like good expected behavior from the user perspective. I just want to make sure that the `EcmascriptEngine` API allows evaluating an emptry string since it is technically a valid operation... but that doesn't mean ReactAppRoot needs to let an empty bundle file go through during hot reload